src/mds/Locker.cc

```diff
@@ -4786,12 +4831,70 @@ void Locker::scatter_tick()
 	       << *lock << " " << *lock->get_parent() << dendl;
       continue;
     }
-    if (now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
+
+    if (!mdr && now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
       break;
     updated_scatterlocks.pop_front();
-    scatter_nudge(lock, 0);
+
+    if (lock->get_type() == CEPH_LOCK_INEST)
+      dout(20) << __func__ << " lock: " << *lock << " of " << *lock->get_parent()
+	<< ", update_stamp: " << lock->get_update_stamp() << ", lock's parent's rstat_dirty_from:"
+	<< static_cast<CInode*>(lock->get_parent())->get_rstat_dirty_from() << ", lock's parent's subdir rstat_dirty_from:"
+	<< static_cast<CInode*>(lock->get_parent())->get_dirfrags_rstat_dirty_from()
+	<< ", mdr's rstats_propagate_time: " << (mdr ? mdr->get_op_stamp() : utime_t()) << dendl;
+
+    if (mdr) {
+
+      if (lock->is_nudged_by(mdr->reqid)) {
+	updated_scatterlocks.push_back(lock->get_updated_item());
+	nudged++;
+	// the lock is nudged by myself previously, and there must
+	// be a C_Locker_RstatFlush that will be triggered, so just
+	// skip this lock for now.
+	continue;
+      }
+
+      CInode* ino = static_cast<CInode*>(lock->get_parent());
+
+      if (lock->get_type() == CEPH_LOCK_INEST
+	  && ino->need_to_nudge(mdr->in[0], mdr->get_op_stamp())) {
+        dout(20) << __func__ << " " << (mdr ? mdr->reqid : metareqid_t())
+          << " need to nudge scatterlock: " << *lock << " of " << *lock->get_parent() << dendl;
+
+        scatter_nudge(lock, new C_Locker_RstatFlush(this, mdr, lock));
+	lock->add_nudged_by(mdr->reqid);
+	nudged++;
+      } else
+        scatter_nudge(lock, 0);
+    } else {
+      scatter_nudge(lock, 0);
+    }
   }
+
   mds->mdlog->flush();
+
+  dout(20) << __func__ << " rstat flush: nudged: " << nudged << dendl;
+
+  // If no locks is of the current mdr's interest, all rstats are flushed successfully
+  if (mdr && !nudged) {
+    dout(20) << __func__ << " " << mdr->reqid
+      << " rstat flush finished!" << dendl;
+
+    // If the inode is projected and we reply clients a success ack,
+    // there exists chances that later getattrs, who comes before
+    // the projected inode poped and retrieves the non-projected
+    // inode attrs, get an out-dated rstat.
+    //
+    // So we wait until the current projected_inode is poped to ack
+    // the clients.
+    if (mdr->in[0]->is_projected() && mdr->client_request && !mdr->already_replied) {
+      mdr->in[0]->add_waiter(CInode::WAIT_POPPROJECTED,
+	  new C_Locker_RespondToRequest(this, mdr));
+      mdr->already_replied = true;
```

Comment:
you'd better to use mdlog->wait_for_safe()

Reply:
@ukernel Hi, I think we have to make sure clients get the response after the projected inode is poped, if we use "mdlog->wait_for_safe()" there's still a possibility that the C_Locker_RespondToRequest is invoked before the Context that pops the projected inode, in which case there exists a chance that the clients' later getattr retrieves an out-dated rstat. Am I right? Thanks:-)